File: server/src/main/java/gg/mineral/server/config/GroovyConfig.java
package gg.mineral.server.config;

import java.io.File;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import dev.zerite.craftlib.chat.component.BaseChatComponent;
import dev.zerite.craftlib.chat.component.StringChatComponent;
import dev.zerite.craftlib.chat.type.ChatColor;
import groovy.util.ConfigObject;
import groovy.util.ConfigSlurper;
import lombok.Data;
import lombok.SneakyThrows;
import lombok.val;

@Data
public class GroovyConfig {
    @Setting("server")
    private String brandName = "Mineral";
    @Setting("server")
    private String motd = ChatColor.BLUE + "Custom Minecraft Server Software";
    @Setting("server")
    private int maxPlayers = 2024;
    @Setting("server")
    private List<String> whitelistedPlayers = Collections.emptyList();
    @Setting("server")
    private String whitelistMessage = ChatColor.RED + "You are not whitelisted on this server.";
    @Setting("server")
    private String serverFullMessage = ChatColor.RED + "The server is full.";
    @Setting("server")
    private String outdatedClientMessage = ChatColor.RED + "Outdated client.";
    @Setting("server")
    private String outdatedServerMessage = ChatColor.RED + "Outdated server.";
    @Setting("server")
    private String restartMessage = ChatColor.RED + "Server is restarting.";
    @Setting("server")
    private String shutdownMessage = ChatColor.RED + "Server closed.";
    @Setting("server")
    private final BaseChatComponent disconnectUnknown = new StringChatComponent(
            ChatColor.RED + "You have been disconnected from the server.");
    @Setting("server")
    private final BaseChatComponent disconnectCanNotAuthenticate = new StringChatComponent(
            ChatColor.RED + "Your account was unable to be authenticated.");
    @Setting("server")
    private final BaseChatComponent disconnectAlreadyLoggedIn = new StringChatComponent(
            ChatColor.RED + "Your account was already logged in.");
    @Setting("player")
    private int viewDistance = 10;
    @Setting("player")
    private int relativeMoveFrequency = 2;
    @Setting("player.hunger")
    private float blockBreakExhaustion = 0.025f;
    @Setting("player.hunger")
    private float swimmingExhaustion = 0.015f;
    @Setting("player.hunger")
    private float walkExhaustion = 0.2f;
    @Setting("player.hunger")
    private float sprintExhaustion = 0.8f;
    @Setting("player.hunger")
    private float combatExhaustion = 0.3f;
    @Setting("player.hunger")
    private float regenExhaustion = 3.0f;
    @Setting("combat")
    private float playerBlockingDamageMultiplier = 0.5f;
    @Setting("combat")
    private boolean disablePlayerCrits = false;
    @Setting("combat.backtrack")
    private boolean backtrackRandom = false;
    @Setting("combat.backtrack")
    private boolean backtrackEnabled = false;
    @Setting("combat.backtrack")
    private boolean comboMode = false;
    @Setting("combat.backtrack")
    private double delayDistanceMin; // in blocks
    @Setting("combat.backtrack")
    private double delayDistanceMax; // in blocks
    @Setting("combat.backtrack")
    private int delayFactor; // in milliseconds
    @Setting("combat.backtrack")
    private int decayFactor; // in milliseconds
    @Setting("combat.backtrack")
    private int maxDelayMs; // in milliseconds
    @Setting("combat.backtrack")
    private int delayResetTime; // in ticks (20 ticks per second)
    @Setting("combat.backtrack")
    private int rMin, rMax;
    @Setting("entity")
    private boolean disableExplosionKnockback = false;
    @Setting("entity")
    private boolean disablePearlKnockback = false;
    @Setting("connection")
    private boolean onlineMode = true;
    @Setting("connection")
    private boolean proxy = false;// TODO: proxy
    @Setting("connection")
    private int port = 25565;
    @Setting("connection")
    private int networkCompressionThreshold = 256; // TODO: network compression
    @Setting("connection.packet-limiter") // TODO: packet limiter
    private String kickMessage = "Too many packets sent!";
    @Setting("plugin")
    private List<String> hiddenPlugins = Arrays.asList("SpookyAC");
    @Setting("plugin")
    private String pluginsFolder = "plugins";
    @Setting("world")
    private String worldsFolder = "worlds";

    private final File configFile = new File("config.groovy");

    @SneakyThrows
    public void load() {
        val configSlurper = new ConfigSlurper();

        if (!configFile.exists())
            configFile.createNewFile();

        val config = configSlurper.parse(configFile.toURI().toURL());

        for (val field : this.getClass().getDeclaredFields()) {
            val setting = field.getDeclaredAnnotation(Setting.class);
            if (setting == null)
                continue;

            val keys = setting.value().split("\\.");
            ConfigObject subConfig = null, lastSubConfig = config;

            for (val key : keys) {
                val nextObject = (ConfigObject) lastSubConfig.get(key);

                if (nextObject == null)
                    break;

                subConfig = nextObject;
                lastSubConfig = subConfig;
            }

            if (subConfig == null)
                continue;

            val value = subConfig.get(field.getName());

            if (value == null)
                continue;

            field.set(this, value);
        }
    }
}

File: server/src/main/kotlin/gg/mineral/server/config/GroovyConfig.kt
package gg.mineral.server.config

import dev.zerite.craftlib.chat.component.BaseChatComponent
import dev.zerite.craftlib.chat.component.StringChatComponent
import dev.zerite.craftlib.chat.type.ChatColor
import groovy.util.ConfigObject
import groovy.util.ConfigSlurper
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import java.io.File

class GroovyConfig {
    @Setting("server")
    val disconnectUnknown: BaseChatComponent = StringChatComponent(
        ChatColor.RED.toString() + "You have been disconnected from the server."
    )
    val configFile: File = File("config.groovy")

    @Setting("server")
    val disconnectCanNotAuthenticate: BaseChatComponent = StringChatComponent(
        ChatColor.RED.toString() + "Your account was unable to be authenticated."
    )

    @Setting("server")
    val disconnectAlreadyLoggedIn: BaseChatComponent = StringChatComponent(
        ChatColor.RED.toString() + "Your account was already logged in."
    )

    @Setting("plugin")
    var pluginsFolder: String = "plugins"

    @Setting("connection")
    var port: Int = 25565

    @Setting("world")
    var worldsFolder: String = "worlds"

    @Setting("server")
    val brandName = "Mineral"

    @Setting("server")
    val motd = ChatColor.BLUE.toString() + "Custom Minecraft Server Software"

    @Setting("server")
    val maxPlayers = 2024

    @Setting("server")
    val whitelistedPlayers = emptyList<String>()

    @Setting("server")
    val whitelistMessage = ChatColor.RED.toString() + "You are not whitelisted on this server."

    @Setting("server")
    val serverFullMessage = ChatColor.RED.toString() + "The server is full."

    @Setting("server")
    val outdatedClientMessage = ChatColor.RED.toString() + "Outdated client."

    @Setting("server")
    val outdatedServerMessage = ChatColor.RED.toString() + "Outdated server."

    @Setting("server")
    val restartMessage = ChatColor.RED.toString() + "Server is restarting."

    @Setting("server")
    val shutdownMessage = ChatColor.RED.toString() + "Server closed."

    @Setting("player")
    val viewDistance = 10

    @Setting("player")
    val relativeMoveFrequency = 2

    @Setting("player.hunger")
    val blockBreakExhaustion = 0.025f

    @Setting("player.hunger")
    val swimmingExhaustion = 0.015f

    @Setting("player.hunger")
    val walkExhaustion = 0.2f

    @Setting("player.hunger")
    val sprintExhaustion = 0.8f

    @Setting("player.hunger")
    val combatExhaustion = 0.3f

    @Setting("player.hunger")
    val regenExhaustion = 3.0f

    @Setting("combat")
    val playerBlockingDamageMultiplier = 0.5f

    @Setting("combat")
    val disablePlayerCrits = false

    @Setting("combat.backtrack")
    val backtrackRandom = false

    @Setting("combat.backtrack")
    val backtrackEnabled = false

    @Setting("combat.backtrack")
    val comboMode = false

    @Setting("combat.backtrack")
    val delayDistanceMin = 0.0 // in blocks

    @Setting("combat.backtrack")
    val delayDistanceMax = 0.0 // in blocks

    @Setting("combat.backtrack")
    val delayFactor = 0 // in milliseconds

    @Setting("combat.backtrack")
    val decayFactor = 0 // in milliseconds

    @Setting("combat.backtrack")
    val maxDelayMs = 0 // in milliseconds

    @Setting("combat.backtrack")
    val delayResetTime = 0 // in ticks (20 ticks per second)

    @Setting("combat.backtrack")
    val rMin = 0

    @Setting("combat.backtrack")
    val rMax = 0

    @Setting("entity")
    val disableExplosionKnockback = false

    @Setting("entity")
    val disablePearlKnockback = false

    @Setting("connection")
    val onlineMode = true

    @Setting("connection")
    val proxy = false // TODO: proxy

    @Setting("connection")
    val networkCompressionThreshold = 256 // TODO: network compression

    @Setting("connection.packet-limiter") // TODO: packet limiter
    val kickMessage = "Too many packets sent!"

    @Setting("plugin")
    val hiddenPlugins = listOf("SpookyAC")

    fun load() {
        val configSlurper = ConfigSlurper()

        if (!configFile.exists()) configFile.createNewFile()

        val config = configSlurper.parse(configFile.toURI().toURL())
        val configCache = Object2ObjectOpenHashMap<String, Any>()

        flattenConfig("", config, configCache)

        for (field in javaClass.declaredFields) {
            val setting = field.getDeclaredAnnotation(
                Setting::class.java
            )
            if (setting == null) continue

            val key = setting.value + "." + field.name
            val value = configCache[key] ?: continue

            field.isAccessible = true
            field[this] = value
        }
    }

    private fun flattenConfig(prefix: String, source: ConfigObject, target: Object2ObjectOpenHashMap<String, Any>) {
        source.forEach { (key: Any?, value: Any?) ->
            val newKey = if (prefix.isEmpty()) key.toString() else prefix + "." + key.toString()
            if (value is ConfigObject) flattenConfig(newKey, value, target)
            else target[newKey] = value
        }
    }
}
